### ISSUE TITLE
FMI2 call sequence bug | fmi2SetInteger/fmi2SetBoolean/fmi2SetString never called while simulation is running

### ISSUE THREAD

**Issue by nik5346** [OpenModelica/OpenModelica#10915]
### Description
When simulating FMUs, currently only fmi2SetReal is called of the four fmi2SetXXX functions during simulation. The fmi2SetInteger, fmi2SetBoolean and fmi2SetString functions are not called during simulation under any circumstances (neither in continuous time mode nor in event mode).

In the FMI standard documentation (version 2.0.4, page 90) the functions fmi2SetReal, fmi2SetInteger, fmi2SetBoolean and fmi2SetString are allowed to be called in event mode. In continuous time mode, however, only the call to fm2SetReal is allowed. Strictly speaking, the bug described above should be fixed so that fmi2SetInteger, fmi2SetBoolean and fmi2SetString are called in event mode. Since there is no obvious reason to not call these three functions in continuous time mode, this bug can also be fixed by simply calling the three functions in continuous time mode. The second solution is not FMI2 standard compliant, but this approach is consistent with other simulation tools (Matlab-Simulink (https://de.mathworks.com/help/simulink/ug/fmu-block-calling-sequences.html), Dymola and MapleSim). In my opinion, the second outlined approach is the best way to resolve the bug.

The described bug can be reproduced with a minimal FMU (https://github.com/nik5346/FeedThroughFmu.git). This FMU has three inputs and three outputs. The values of the inputs are simply passed on to the outputs within the FMU. The first input-output pair is of type real, the second pair is of type integer and the last pair is of type bool. In addition, this FMU periodically triggers the event mode by a timed event. Simulating this FMU in OpenModelica and connecting step signals to the inputs, only the input signal at the input of type real is passed to the corresponding output. The other two outputs remain constant.

The call sequence of the current OpenModelica build looks like this:
...
fmi2CompletedIntegratorStep
fmi2GetReal
fmi2GetInteger
fmi2GetBoolean
fmi2SetTime
fmi2SetReal
fmi2CompletedIntegratorStep
fmi2EnterEventMode
fmi2NewDiscreteStates
fmi2EnterContinuousTimeMode
fmi2GetReal
fmi2GetInteger
fmi2GetBoolean
fmi2SetTime
fmi2SetReal
fmi2CompletedIntegratorStep
fmi2GetReal
...

The simulation result of the current OpenModelica build looks like this:
![image](https://github.com/OpenModelica/OpenModelica/assets/87228885/095fcf3d-5e60-4526-9d06-55be3f459456)

### Expected behavior
With my changes (#10916) the call sequence looks like this:
...
fmi2CompletedIntegratorStep
fmi2GetReal
fmi2GetInteger
fmi2GetBoolean
fmi2SetTime
fmi2SetReal
fmi2SetInteger
fmi2SetBoolean
fmi2CompletedIntegratorStep
fmi2EnterEventMode
fmi2NewDiscreteStates
fmi2EnterContinuousTimeMode
fmi2GetReal
fmi2GetInteger
fmi2GetBoolean
fmi2SetTime
fmi2SetReal
fmi2SetInteger
fmi2SetBoolean
fmi2CompletedIntegratorStep
...

The simulation result with my changes looks like this:
![image](https://github.com/OpenModelica/OpenModelica/assets/87228885/6ef7d825-2bfc-4b42-836f-7d4ad58d5ebb)

**Comment by casella** [comment]
@nik5346 it is not clear what you mean by "When simulating FMUs,". How are you running the simulations: with OMSimulator or by importing the FMU in a Modelica model and then using OpenModelica to run it? How do you generate the FMU? Can you post an example to reproduce the issue?

Thanks!

**Comment by nik5346** [comment]
@casella 
I imported the FMU into an otherwise empty model in OpenModelica.
The model looks like this:
![image](https://github.com/OpenModelica/OpenModelica/assets/87228885/9f1b00b6-23f0-4b48-a162-aef25c6e174f)

As already stated above, I created a minimal FMU to reproduce the issue. This is the code I used: https://github.com/nik5346/FeedThroughFmu
You would need to create an FMU from this code your self to reproduce the issue. Just compile it and package the fmu.
There is no generation tool involved.

**Comment by AnHeuermann** [comment]
I reproduced the issue with a 2.0 ME FMU created from

```modelica
model FeedThrough
  input Real x_in;
  output Real x_out = x_in;
  input Integer i_in;
  output Integer i_out = i_in;
  input Boolean b_in;
  output Boolean b_out = b_in;
end FeedThrough;
```

Apparently the event handling in the FMU import in OMEdit is not correct.
The old FMU import is broken in many places and I would recommend to use SSP for connecting FMUs. It usually works way better.

![grafik](https://github.com/OpenModelica/OpenModelica/assets/38031952/7e6e5a92-7060-4229-b868-72598baec293)

But in this case SSP seems to break as well, not sure why.

<details>
<summary>Error message</summary>

```
LOG_INIT          | error   | The initialization problem is inconsistent due to the following equation: 0 != -1.5 = x
info:    Result file: TestFeedThrough_res.mat (bufferSize=10)
error:   [fmi2logger] FeedThrough (logStatusError): fmi2SetInteger: Illegal model exchange call sequence. fmi2SetInteger is not allowed in model_state_me_continuous_time_mode state.
error:   [fmi2logger] FeedThrough (logStatusError): fmi2SetTime: Illegal model exchange call sequence. fmi2SetTime is not allowed in model_state_error state.
error:   [doStep] fmi2_import_set_time failed for FMU "TestFeedThrough.Root.FeedThrough"
error:   [fmi2logger] FeedThrough (logStatusError): fmi2CompletedIntegratorStep: Illegal model exchange call sequence. fmi2CompletedIntegratorStep is not allowed in model_state_error state.
error:   [doStep] fmi2_import_completed_integrator_step failed for FMU "TestFeedThrough.Root.FeedThrough"
Traceback (most recent call last):
  File "/home/aheuermann/workspace/OM/OpenModelica/install_cmake/bin/../share/OMSimulator/scripts/OMSimulatorServer.py", line 191, in <module>
    _main()
  File "/home/aheuermann/workspace/OM/OpenModelica/install_cmake/bin/../share/OMSimulator/scripts/OMSimulatorServer.py", line 188, in _main
    server.run()
  File "/home/aheuermann/workspace/OM/OpenModelica/install_cmake/bin/../share/OMSimulator/scripts/OMSimulatorServer.py", line 158, in run
    self._model.doStep()
  File "/home/aheuermann/.local/lib/python3.10/site-packages/OMSimulator/Model.py", line 50, in doStep
    raise Exception('error {}'.format(Types.Status(status)))
Exception: error Status.ERROR
error:   [fmi2logger] FeedThrough (logStatusError): fmi2Terminate: Illegal model exchange call sequence. fmi2Terminate is not allowed in model_state_error state.
error:   [terminate] Termination of system "FeedThrough" failed
error:   [terminate] Termination of system "TestFeedThrough.Root" failed
info:    0 warnings
info:    8 errors
Simulation process failed. Exited with code 1.
```
</details>

**Comment by AnHeuermann** [comment]
@nik5346 if you want to fix it you probably need to check `fmi2EventUpdate_OMC`. But event handling in FMI is complicated and in my opinion we should remove all of the FMI import that doesn't use SSP.

**Comment by mahge** [comment]
@AnHeuermann is it possible that the issue is in the generated FMUs themselves? I am currently looking at some failing fmu-tests on the [fmi-master](https://libraries.openmodelica.org/branches/master-fmi/Modelica_4.0.0/Modelica_4.0.0.html) regression testing (which uses OMSimulator) and I feel like there is some issue with event handling. For example, `Modelica.Blocks.Examples.Filter` we get some signals failing to match the expected curve. 

This is what the graph looks like for the [`Bessel.x[1]`](https://[libraries.openmodelica.org/branches/master-fmi/Modelica_4.0.0/files/Modelica_4.0.0_Modelica.Blocks.Examples.Filter.diff.Bessel.x[1].html](https://libraries.openmodelica.org/branches/master-fmi/Modelica_4.0.0/files/Modelica_4.0.0_Modelica.Blocks.Examples.Filter.diff.Bessel.x[1].html)) signal of the model, for example (the blue curve is the expected value and the red curve is the actual value)

![image](https://github.com/OpenModelica/OpenModelica/assets/13765810/c8e9bcda-7f6e-4033-9ed9-828d13748def)


I am not sure how to interpret it yet, but in layman terms, it seems like the event response in the FMU is delayed a bit.  I am investigating and hopefully we will find out why.

**Comment by nik5346** [comment]
@mahge 
In FMI2ModelExchange.c the FMISolvingMode is set to fmi2_event_mode only in two occasions.
First one is here:
```cpp
void fmi2ExitInitializationModel_OMC(void* in_fmi2me)
{
  FMI2ModelExchange* FMI2ME = (FMI2ModelExchange*)in_fmi2me;
  fmi2_status_t status = fmi2_import_exit_initialization_mode(FMI2ME->FMIImportInstance);
  FMI2ME->FMISolvingMode = fmi2_event_mode;
  if (status != fmi2_status_ok && status != fmi2_status_warning) {
    ModelicaFormatError("fmi2ExitInitializationMode failed with status : %s\n", fmi2_status_to_string(status));
  }
}
```
Second one is here:
```cpp
int fmi2EventUpdate_OMC(void* in_fmi2me)
{
  FMI2ModelExchange* FMI2ME = (FMI2ModelExchange*)in_fmi2me;
  fmi2_event_info_t *eventInfo = FMI2ME->FMIEventInfo;
  fmi2_status_t status = fmi2_import_enter_event_mode(FMI2ME->FMIImportInstance);
  if (status != fmi2_status_ok && status != fmi2_status_warning) {
    ModelicaFormatError("fmi2EnterEventMode failed with status : %s\n", fmi2_status_to_string(status));
  }
  FMI2ME->FMISolvingMode = fmi2_event_mode;

  eventInfo->newDiscreteStatesNeeded = fmi2_true;
  eventInfo->terminateSimulation = fmi2_false;
  status = fmi2_import_new_discrete_states(FMI2ME->FMIImportInstance, eventInfo);
  if (status != fmi2_status_ok && status != fmi2_status_warning) {
    ModelicaFormatError("fmi2NewDiscreteStates failed with status : %s\n", fmi2_status_to_string(status));
  }

  status = fmi2_import_enter_continuous_time_mode(FMI2ME->FMIImportInstance);
  if (status != fmi2_status_ok && status != fmi2_status_warning) {
    ModelicaFormatError("fmi2EnterContinuousTimeMode failed with status : %s\n", fmi2_status_to_string(status));
  }
  FMI2ME->FMISolvingMode = fmi2_continuousTime_mode;
  return eventInfo->valuesOfContinuousStatesChanged;
}
```

fmi2ExitInitializationModel_OMC is usually only called once at the beginning of the simulation and therefor irrelevant for this issue. fmi2EventUpdate_OMC is called when ever the fmu is in event mode. As you can see, this function sets FMISolvingMode to fmi2_event_mode. Discret states get updated and then immediately FMISolvingMode is set to fmi2_continuousTime_mode. Whenever fmi2SetInteger_OMC is called later, it always fails to set the variables, since it checks the FMISolvingMode to be fmi2_event_mode.

```cpp
void fmi2SetInteger_OMC(void* in_fmi2, int numberOfValueReferences, double* integerValuesReferences, int* integerValues, int fmiType)
{
  if (fmiType == 1) {
    FMI2ModelExchange* FMI2ME = (FMI2ModelExchange*)in_fmi2;
    if (FMI2ME->FMISolvingMode == fmi2_instantiated_mode || FMI2ME->FMISolvingMode == fmi2_initialization_mode || FMI2ME->FMISolvingMode == fmi2_event_mode /*|| FMI2ME->FMISolvingMode == fmi2_continuousTime_mode*/) {
      fmi2_value_reference_t* valuesReferences_int = real_to_fmi2_value_reference(numberOfValueReferences, integerValuesReferences);
      fmi2_status_t status = fmi2_import_set_integer(FMI2ME->FMIImportInstance, valuesReferences_int, numberOfValueReferences, (fmi2_integer_t*)integerValues);
      free(valuesReferences_int);
      if (status != fmi2_status_ok && status != fmi2_status_warning) {
        ModelicaFormatError("fmi2SetInteger failed with status : %s\n", fmi2_status_to_string(status));
      }
    }
  } else if (fmiType == 2) {

  }
}
```

Therefor the described issue is definitely a call sequence bug. Moreover I provided code for a self composed minimal FMU which works fine in all Simulation-Tools, but not in OpenModelica.

**Comment by casella** [comment]
FYI, FMU import is still problematic, also because it is not clear how to do that in general. There is some work going on within the Modelica Association that may eventually lead to a nice solution to this issue. It is likely that we will test it later this year.

Please follow #5345 for updates.

**Comment by nik5346** [comment]
I fixed the issue again in #10943.

The fmi2SetInteger, fmi2SetBoolean and fmi2SetString functions are now only called in event mode. This solution does not violate the fmi2 standard.

The call sequence looks like this:
...
fmi2CompletedIntegratorStep
fmi2GetReal
fmi2GetInteger
fmi2GetBoolean
fmi2SetTime
fmi2SetReal
fmi2CompletedIntegratorStep
fmi2EnterEventMode
fmi2SetReal
fmi2SetInteger
fmi2SetBoolean
fmi2NewDiscreteStates
fmi2EnterContinuousTimeMode
fmi2GetReal
fmi2GetInteger
fmi2GetBoolean
fmi2SetTime
fmi2SetReal
fmi2CompletedIntegratorStep
...

My self composed feedthrough FMU works fine.

**Comment by casella** [comment]
@nik5346 you need to make sure that your PR doesn't break the testsuite, otherwise it cannot be merged on master.

**Comment by nik5346** [comment]
@casella Thanks for the advice. I did not notice the test fail. However I updated the pull request. The tests succeed now.

**Comment by nik5346** [comment]
I requested a clarification on when to call the fmi2SetXXX functions.
https://github.com/modelica/fmi-standard/issues/1890

**Comment by casella** [comment]
@nik5346 I couldn't follow the discussion in detail because I'm not too much into the FMI technicalities. As far as I understand, the rationale is that continuous Real variables v(t) conceptually represent function of Real time, so their value can change between any two Real time variables, continuously or discontinuously (in which case you need an event to do things properly). Of course the Real time is then approximated with IEEE 754 double precision, which is kind of discrete, but you pretend it's actually continouous. Anyway, that's another story.

Discrete variables instead only change at isolated instants (time or state events). They cannot change continuously over time, because their values are discrete. So, they are defined as discrete by default.

In your case, the Integer variable is also an input, which makes things a bit more complicated, but the fundamentals still apply. In the other thread you state:

> Consider a feedthrough system (input directly maps to output). Both input and output are of type integer.
In order to simulate this model correctly, you would need to trigger an event at every time instant in order to get the simulation environment to call fmi2SetInteger. This is a design flaw in my opinion.

If you want to do proper error control, in principle you'd have to trigger an event each time, because the value could change abruptly at some time instant which is not known a priori, so the thing should be handled properly with event detection and stuff. Of course it also depends where the input comes from, if it comes from another FMI, that's one thing, if it comes from a socket or data file, it's another thing. But I guess the standard cannot depend on that.

Of course a smart implementation of the FMI and of the solver handling it could check if the integer input has changed, and if it hasn't, it could avoid the event processing entirely, and continue with undisturbed continuous-time integration. That's a quality-of-implementation issue.

If you want to avoid that entirely, a simple workaround could be to convert those Integer input into Real, and then do whatever best fits you inside the FMU to get back to Integer.

If other tools allow changing integer inputs outside events, then they are positively not conforming to the standard, and you should complain with the tool vendors, not with the working group that develops the standard.

Are you really sure that this restriction is lifted in FMI 3.0? Where do you see that in the specification?

My 2 cts 😅

**Comment by nik5346** [comment]
@casella 
> Are you really sure that this restriction is lifted in FMI 3.0? Where do you see that in the specification?

They did actually not lift this restriction. The FMI3 documentation is similarly misleading on that matter as the FMI2 documentation was.

I am ok, when you want the FMI import to be standard compliant. So PR #10945 it is then?

At this point in time there is still not way to trigger the fmi2SetInteger, fmi2SetBoolean and fmi2SetString functions. Therefor OpenModelica is not standard compliant at the moment. This bug still needs a fix.

PR #10945 is absolutely standard compliant and the only correct implementation, as clarified in https://github.com/modelica/fmi-standard/issues/1890.
On the other hand PR #10945 still does not allow you to simulate a feedthrough FMU exported from other tools, because their FMUs don't trigger the event mode.

It's your choice not mine. I can deal with both proposed solutions from within my FMU. I definitely need a fix for this bug otherwise, I would need to recommend to other not to use OpenModelica with my FMU, since OpenModelica has a disfunctional FMU import. I don't really want this, thats why I created the PRs.

**Comment by casella** [comment]
@nik5346, I briefly reviewed this ticket and the two PRSs. To be honest, I am a bit confused, because they mix up a critique of the FMI standard, issues with OMC and issues with other tools. The critique of the FMI standard may be OK, but this is not the appropriate place to discuss it.

May I suggest that you open a new clean issue where you do not question the current existing FMI standard, nor involve the behaviour of other tools, and only describe what you think should be fixed in OMC to make it compliant to the standard? IMHO that would help speeding up this issue.

Thanks!

**Comment by nik5346** [comment]
> FYI, there is a proposal coming up to actually define how FMI import into a Modelica model should be done. The proposal includes open-source code to actually generate the Modelica code of the imported FMU.

I like the idea of having a open-source solution for generating the Modelica code. This may reduce my complains about the fmi standard in general. 

> May I suggest that you open a new clean issue where you do not question the current existing FMI standard, nor involve the behaviour of other tools, and only describe what you think should be fixed in OMC to make it compliant to the standard currently in force? IMHO that would help speeding up this issue.

I agree, adding the behavior of other tools to the discussion increased the confusion. Let's discuss the desired simulation result and required calling sequence only based of the fmi standard itself. I create a new issue.

**Comment by nik5346** [comment]
Here it is: #10985
